widgets/src/lib/InferenceWidget/widgets/ImageSegmentationWidget/ImageSegmentationWidget.svelte

```diff
@@ -0,0 +1,282 @@
+						.map((o) => addOutputCanvasData(o))
+				);
+				isLoading = false;
+			}
+			// outputJson = res.outputJson;
+		} else if (res.status === "loading-model") {
+			modelLoading = {
+				isLoading: true,
+				estimatedTime: res.estimatedTime,
+			};
+			getOutput(file, true);
+		} else if (res.status === "error") {
+			error = res.error;
+		}
+	}
+
+	function isValidOutput(arg: any): arg is ImageSegment[] {
+		return (
+			Array.isArray(arg) &&
+			arg.every(
+				(x) =>
+					typeof x.label === "string" &&
+					typeof x.score === "number" &&
+					typeof x.mask === "string"
+			)
+		);
+	}
+	function parseOutput(body: unknown): ImageSegment[] {
+		if (isValidOutput(body)) {
+			return body;
+		}
+		throw new TypeError(
+			"Invalid output: output must be of type Array<{label:string; score:number; mask: string}>"
+		);
+	}
+
+	function mouseout() {
+		highlightIndex = -1;
+	}
+
+	function mouseover(index: number) {
+		highlightIndex = index;
+	}
+
+	function mousemove(e: any, canvasW: number, canvasH: number) {
+		let { layerX, layerY } = e;
+		layerX = clamp(layerX, 0, canvasW);
+		layerY = clamp(layerY, 0, canvasH);
+		const row = Math.floor((layerX / canvasH) * imgH);
+		const col = Math.floor((layerY / canvasW) * imgW);
+		highlightIndex = -1;
+		const index = (imgW * col + row) * 4;
+		for (const [i, o] of output.entries()) {
+			const pixel = o.imgData.data[index];
+			if (pixel > 0) {
+				highlightIndex = i;
+			}
+		}
+	}
+
+	function addOutputColor(imgSegment: ImageSegment, idx: number) {
+		const hash = mod(idx, COLORS.length);
+		const { color } = COLORS[hash];
+		return { ...imgSegment, color };
+	}
+
+	async function addOutputCanvasData(
+		imgSegment: ImageSegment
+	): Promise<ImageSegment> {
+		const { mask, color } = imgSegment;
+
+		const maskImg = new Image();
+		maskImg.src = `data:image/png;base64, ${mask}`;
+		// await image.onload
+		await new Promise((resolve, _) => {
+			maskImg.onload = () => resolve(maskImg);
+		});
+		const imgData = getImageData(maskImg);
+		const { r, g, b } = colorToRgb[color];
+		const maskColored = [r, g, b, maskOpacity];
+		const background = Array(4).fill(0);
+
+		for (let i = 0; i < imgData.data.length; i += 4) {
+			const [r, g, b, a] = imgData.data[i] === 255 ? maskColored : background;
+			imgData.data[i] = r;
+			imgData.data[i + 1] = g;
+			imgData.data[i + 2] = b;
+			imgData.data[i + 3] = a;
+		}
+
+		const bitmap = await createImageBitmap(imgData);
+		return { ...imgSegment, imgData, bitmap };
+	}
+
+	function getImageData(maskImg: CanvasImageSource): ImageData {
+		const tmpCanvas = document.createElement("canvas");
+		tmpCanvas.width = imgW;
+		tmpCanvas.height = imgH;
+		const tmpCtx = tmpCanvas.getContext("2d");
+		tmpCtx.drawImage(maskImg, 0, 0, imgW, imgH);
+		const segmentData = tmpCtx.getImageData(0, 0, imgW, imgH);
+		return segmentData;
+	}
+
+	// original: https://gist.github.com/MonsieurV/fb640c29084c171b4444184858a91bc7
+	function polyfillCreateImageBitmap() {
```

Comment:
> we may want to put this in a shared utils function ?

I'll keep this polyfill in ImageSegmentationWidget.svelte since this polyfill fills only specific subset of createImageBitmap API that is only needed for this widget. 

> And don't forget to revoke the blobURIs, depending from where they come from you are possibly creating a memory leak.

I've updated the changes in the polyfill in[ this commit](https://github.com/huggingface/huggingface_hub/pull/378/commits/2f4ea5cd6227ae84d38e62248403ca8c5d4804c7). Removed the line: `dataURL = URL.createObjectURL(data);`, which was not used in ImgSegWidget and would have been a memory concern. Moreover, I've tested with Safari dev tools to make sure that garbage collection was correctly freeing memories being used in Polyfills

Reply:
Outstanding 🙏